zerver/tests/test_subs.py

```diff
@@ -2984,7 +2984,7 @@ def test_multi_user_subscription(self) -> None:
                     streams_to_sub,
                     dict(principals=orjson.dumps([user1.id, user2.id]).decode()),
                 )
-        self.assert_length(queries, 35)
+        self.assert_length(queries, 34)
```

Comment:
Do you understand why these query counts went down?  One theory I had is these hunks belong in the last commit, and the i18n changes reduced a commit.

Reply:
The test_subs query count decreases appear to be due to this query missing :
```
{'sql': 'SELECT "django_session"."session_key", "django_session"."session_data", "django_session"."expire_date" FROM "django_session" WHERE ("django_session"."expire_date" > \'2021-01-20T13:42:18.092543+00:00\'::timestamptz AND "django_session"."session_key" = \'xavzuf6q21cinvkc65ya91bbcy6t0q44\')', 'time': '0.001'}
```

Not completely sure why, but it does seem like due to i18n changes - since Django stops looking for language in the session variables in 3.0, when handling these requests it probably no longer needs to look at any session variables and thus has no need to retrieve session data from the database.

test_users query decrease by 2 is due to Django optimizations - in `do_create_user` codepath we enqueue 2 Scheduled emails. Old Django does these queries:
```
{'sql': 'INSERT INTO "zerver_scheduledemail" ("scheduled_timestamp", "data", "realm_id", "address", "type") VALUES (\'2021-01-20T16:10:48.378071+00:00\'::timestamptz, \'{"template_prefix":"zerver/emails/followup_day1","from_name":null,"from_address":"SUPPORT","language":null,"context":{"realm_uri":"http://zulip.testserver","realm_name":"Zulip Dev","root_domain_uri":"http://testserver","external_uri_scheme":"http://","external_host":"testserver","user_name":"Fred Flintstone","unsubscribe_link":"http://zulip.testserver/accounts/unsubscribe/welcome/j6ghrjw3lqb5tmpd22brj3hj","keyboard_shortcuts_link":"http://zulip.testserver/help/keyboard-shortcuts","realm_creation":false,"email":"fred@zulip.com","is_realm_admin":false,"getting_started_link":"https://zulip.com"}}\', 2, NULL, 1) RETURNING "zerver_scheduledemail"."id"', 'time': '0.000'}
{'sql': 'SELECT "zerver_scheduledemail_users"."userprofile_id" FROM "zerver_scheduledemail_users" WHERE ("zerver_scheduledemail_users"."scheduledemail_id" = 2 AND "zerver_scheduledemail_users"."userprofile_id" IN (24))', 'time': '0.000'}
{'sql': 'INSERT INTO "zerver_scheduledemail_users" ("scheduledemail_id", "userprofile_id") VALUES (2, 24) RETURNING "zerver_scheduledemail_users"."id"', 'time': '0.001'}
```

Django 3.0 does:
```
{'sql': 'INSERT INTO "zerver_scheduledemail" ("scheduled_timestamp", "data", "realm_id", "address", "type") VALUES (\'2021-01-20T16:15:17.361726+00:00\'::timestamptz, \'{"template_prefix":"zerver/emails/followup_day1","from_name":null,"from_address":"SUPPORT","language":null,"context":{"realm_uri":"http://zulip.testserver","realm_name":"Zulip Dev","root_domain_uri":"http://testserver","external_uri_scheme":"http://","external_host":"testserver","user_name":"Fred Flintstone","unsubscribe_link":"http://zulip.testserver/accounts/unsubscribe/welcome/ptpwpjkabih5v3ldoxaiixwc","keyboard_shortcuts_link":"http://zulip.testserver/help/keyboard-shortcuts","realm_creation":false,"email":"fred@zulip.com","is_realm_admin":false,"getting_started_link":"https://zulip.com"}}\', 2, NULL, 1) RETURNING "zerver_scheduledemail"."id"', 'time': '0.000'}
{'sql': 'INSERT INTO "zerver_scheduledemail_users" ("scheduledemail_id", "userprofile_id") VALUES (2, 24) ON CONFLICT DO NOTHING', 'time': '0.001'}
```

So, by usng `ON CONFLICT DO NOTHING` it's able to eliminate a SELECT in the `email.users.add(*to_user_ids)` line in `send_future_email`. 2 emails are being scheduled, thus query count decrease by two.

I haven't investigated test_signup, but it's registration, so also do_create_user codepath, so the above explanation should apply. The one remaining query count drop from 281 to 251 doesn't seem worth the effort it would take to work through that number of queries.

Reply:
Great, that's very helpful!  It's always nice when your framework upgrade makes things faster :)